synapse/models/infotech.py

```diff
@@ -1,7 +1,143 @@
+        major, minor, patch = s_version.unpackVersion(valu)
+        valu = s_version.fmtVersion(major, minor, patch)
+        return valu
+
+def bruteVersionValu(valu):
+    '''
+    Return the system normalized version integer for a given input.
+
+    Args:
+        valu: String or integer to normalize.
+
+    Returns:
+        int: System normalized version value.
+    '''
+    return bruteVersion(valu)[0]
+
+def bruteVersion(valu):
+    '''
+    Attempt to brute force a valu into a semantic version string and its components
+
+    Args:
+        valu: A string or integer to attempt to obtain a system normalized version
+        valu and subs for.
+
+    Returns:
+        int, dict: The system normalized version integer and a subs dictionary.
+    '''
+    if isinstance(valu, int):
+        return s_datamodel.tlib.getTypeNorm('it:semver', valu)
+
+    if isinstance(valu, str):
+        return bruteStr(valu)
+
+    else:
+        raise s_common.BadTypeValu(valu=valu,
+                                   mesg='Unable to brute force  a valu',
+                                   type=type(valu))
+
+def bruteStr(valu):
+    '''
+    Brute force the version out of a string.
+
+    Args:
+        valu (str): String to attempt to get version information for.
+
+    Notes:
+        This first attempts to parse strings using the it:semver normalization
+        before attempting to extract version parts out of the string.
+
+    Returns:
+        int, dict: The system normalized version integer and a subs dictionary.
+    '''
+    try:
+        valu, subs = s_datamodel.tlib.getTypeNorm('it:semver', valu)
+        return valu, subs
+    except s_common.BadTypeValu:
+        # Try doing version part extraction by noming through the string
+        subs = s_version.parseVersionParts(valu)
+        subs.setdefault('major', 0)
+        subs.setdefault('minor', 0)
+        subs.setdefault('patch', 0)
+        if subs:
+            valu = s_version.packVersion(subs.get('major'),
+                                         subs.get('minor'),
+                                         subs.get('patch'))
+            return valu, subs
+    raise s_common.BadTypeValu(valu=valu,
+                               mesg='Unable to brute force version parts out of the string')  # pragma: no cover
 
 class ItMod(CoreModule):
 
+    def initCoreModule(self):
+        self.onFormNode('it:dev:str', self._onFormItDevStr)
+        self.onFormNode('it:prod:softver', self._onFormItSoftVer)
+        self.core.addTypeCast('it:version:brute', bruteVersionValu)
+
+    def _onFormItDevStr(self, form, valu, props, mesg):
+        props['it:dev:str:norm'] = valu.lower()
+
+    def _onFormItSoftVer(self, form, valu, props, mesg):
+        verstr = props.get('it:prod:softver:verstr')
+        props['it:prod:softver:verstr:norm'] = self.core.getTypeNorm('str:lwr', verstr)[0]
+        if 'it:prod:softver:semver' in props:
+            return
+        valu, subs = bruteVersion(verstr)
```

Comment:
Are we missing a valu is None check here?

Reply:
right now the API for bruting returns `0x0` but its a trivial change to return None and just bail if we can't get any meaninful value for a version string.

Reply:
with the None change in the `version.py` code this will now throw an exception, which will end the callback